# IBM Data Science Capstone
## Segmenting and clustering neighborhoods in Toronto

### Part 1: Transforming the data from Wikipedia to a Pandas DF

In [1]:
# Installing beautiful soup package
! pip install beautifulsoup4
! pip install lxml

In [2]:
# Importing libraries
import pandas as pd
from bs4 import BeautifulSoup
import requests
import numpy as np
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [3]:
# Using Beautiful soop and Pandas to extract the table from the wiki page
res = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
soup = BeautifulSoup(res.content,'html')
table = soup.find_all('table')[0] 
df_neigh = pd.read_html(str(table))[0]
df_neigh

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [4]:
# removing the rows where Borough is NA
index=df_neigh[df_neigh['Borough'] == 'Not assigned'].index
df_neigh.drop(index, inplace=True)
df_neigh.reset_index(inplace=True,drop=True)
df_neigh

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [5]:
# Combine the Neighborhoods that belongs to the same Borough
df_neigh.describe()

Postal Code     Borough Neighbourhood
count          103         103           103
unique         103          10            99
top            M4G  North York     Downsview
freq             1          24             4

#### So, there is 103 unique Postal Codes and 103 rows of data. I think is not necessary to merge nothing here.

In [6]:
index=df_neigh[df_neigh['Neighbourhood'] == 'Not assigned'].index
print(index)

Int64Index([], dtype='int64')


#### Also in this case, the Index function returns an empty list, so, there's no Neighbourhood with status "Not Assigned". Nothing to process here...

In [7]:
df_neigh.shape

(103, 3)

### Part 2: Adding Geo coordinates for each postal code on the DF

#### I've downloaded the file provide by you (thanks!) and converted into a Pandas DF.

In [8]:
df_coord=pd.read_csv("https://cocl.us/Geospatial_data")
df_coord

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

#### OK, we got 103 coordinates for 103 Boroughs. Now, we're going to add the coordinates to the Neighborhoods DF

In [9]:
df_neigh=pd.merge(df_neigh, df_coord, on = "Postal Code")

In [10]:
df_neigh

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

### Part 3: Exploring and clustering Toronto

#### First, we're going to separate in another DF only the Boroughs that have "Toronto" on it's name:

In [11]:
df_neigh['Borough'].unique()

array(['North York', 'Downtown Toronto', 'Etobicoke', 'Scarborough',
       'East York', 'York', 'East Toronto', 'West Toronto',
       'Central Toronto', 'Mississauga'], dtype=object)

In [12]:
df_toronto = df_neigh.query('Borough == "Downtown Toronto" or Borough == "East Toronto" or Borough == "West Toronto" or Borough == "Central Toronto"').reset_index(drop=True)

df_toronto

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

In [13]:
# Import libraries needed
from sklearn.cluster import KMeans
import folium # map rendering library

In [14]:
#Creating the map for Toronto

latitude=43.7184038
longitude=-79.5181405

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Now, we're going to put the Foursquare credentials to explore the neighborhoods

In [15]:
CLIENT_ID = 'PAPMAYIRQ3RDOYDW3OG0WDH055ZOF4GBUN1SPYCHXVMKZZVZ' # your Foursquare ID
CLIENT_SECRET = 'Y3LF1IPE1503EXBQ4C3GS3GPOXZFJJJNSFQ2XRDA2JWGHMIZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: PAPMAYIRQ3RDOYDW3OG0WDH055ZOF4GBUN1SPYCHXVMKZZVZ
CLIENT_SECRET:Y3LF1IPE1503EXBQ4C3GS3GPOXZFJJJNSFQ2XRDA2JWGHMIZ


In [16]:
# use the function of the lab
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


#### Now, we're going to use this function to get the TOP 100 venues in a radius of 500 meters for all the locations in our Toronto DF

In [17]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [18]:
print(toronto_venues.shape)
toronto_venues.head()

(1618, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

#### Let's analize a little bit the data

In [19]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     61   
Christie                                                               15   
Church and Wellesley                                                   80   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           14   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         22   
Kensington Market, Chinatown, Grange Park                              62   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             4   
North Toronto West, Lawrence Park                                      21   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               96   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     37   
St. James Town                                                         81   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         98   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      61   
Christie                                                                15   
Church and Wellesley                                                    80   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [20]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 234 uniques categories.


#### Now, we're going to convert the categorical values to indicator values, for further analysis

In [21]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe, on the first column
# since one of the categories is also 'Neighborhood', we're going to use 'Neigbourhood' instead (as mispelled in the original DF)
# We also could have checked if the column exist and then deleted, but I prefer to maintain all the categories on this point, 
# since we don't know if it's relevant or not
toronto_onehot.insert(0, "Neighbourhood", toronto_venues['Neighborhood'], True)

toronto_onehot.head()

Neighbourhood  Adult Boutique  Afghan Restaurant  Airport  \
0  Regent Park, Harbourfront               0                  0        0   
1  Regent Park, Harbourfront               0                  0        0   
2  Regent Park, Harbourfront               0                  0        0   
3  Regent Park, Harbourfront               0                  0        0   
4  Regent Park, Harbourfront               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Theme Restaurant  \
0                 0                    0  ...                 0   
1                 0                    0  ...                 0   
2                 0                    0  ...                 0   
3                 0                    0  ...                 0   
4                 0                    0  ...                 0   

   Tibetan Restaurant  Toy / Game Store  Trail  Train Station  \
0                   0                 0      0              0   
1                   0                 0      0              0   
2                   0                 0      0              0   
3                   0                 0      0              0   
4                   0                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Yoga Studio  
0         0            0  
1         0            0  
2         0            0  
3         0            0  
4         0            0  

[5 rows x 235 columns]

#### Let's examine the DF

In [22]:
toronto_onehot.shape

(1618, 235)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [23]:
toronto_grouped=toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Adult Boutique  \
0                                         Berczy Park          0.0000   
1        Brockton, Parkdale Village, Exhibition Place          0.0000   
2   Business reply mail Processing Centre, South C...          0.0000   
3   CN Tower, King and Spadina, Railway Lands, Har...          0.0000   
4                                  Central Bay Street          0.0000   
5                                            Christie          0.0000   
6                                Church and Wellesley          0.0125   
7                      Commerce Court, Victoria Hotel          0.0000   
8                                          Davisville          0.0000   
9                                    Davisville North          0.0000   
10                       Dufferin, Dovercourt Village          0.0000   
11             First Canadian Place, Underground city          0.0000   
12    Forest Hill North & West, Forest Hill Road Park          0.0000   
13                           Garden District, Ryerson          0.0000   
14  Harbourfront East, Union Station, Toronto Islands          0.0000   
15                      High Park, The Junction South          0.0000   
16                     India Bazaar, The Beaches West          0.0000   
17          Kensington Market, Chinatown, Grange Park          0.0000   
18                                      Lawrence Park          0.0000   
19                           Little Portugal, Trinity          0.0000   
20                        Moore Park, Summerhill East          0.0000   
21                  North Toronto West, Lawrence Park          0.0000   
22                             Parkdale, Roncesvalles          0.0000   
23        Queen's Park, Ontario Provincial Government          0.0000   
24                          Regent Park, Harbourfront          0.0000   
25                           Richmond, Adelaide, King          0.0000   
26                                           Rosedale          0.0000   
27                                           Roselawn          0.0000   
28                                 Runnymede, Swansea          0.0000   
29                                     St. James Town          0.0000   
30                        St. James Town, Cabbagetown          0.0000   
31                                     Stn A PO Boxes          0.0000   
32                                    Studio District          0.0000   
33  Summerhill West, Rathnelly, South Hill, Forest...          0.0000   
34                The Annex, North Midtown, Yorkville          0.0000   
35                                        The Beaches          0.0000   
36                       The Danforth West, Riverdale          0.0000   
37           Toronto Dominion Centre, Design Exchange          0.0000   
38                     University of Toronto, Harbord          0.0000   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0              0.0000  0.000000            0.000000      0.000000   
1              0.0000  0.000000            0.000000      0.000000   
2              0.0000  0.000000            0.000000      0.000000   
3              0.0000  0.058824            0.058824      0.058824   
4              0.0000  0.000000            0.000000      0.000000   
5              0.0000  0.000000            0.000000      0.000000   
6              0.0125  0.000000            0.000000      0.000000   
7              0.0000  0.000000            0.000000      0.000000   
8              0.0000  0.000000            0.000000      0.000000   
9              0.0000  0.000000            0.000000      0.000000   
10             0.0000  0.000000            0.000000      0.000000   
11             0.0000  0.000000            0.000000      0.000000   
12             0.0000  0.000000            0.000000      0.000000   
13             0.0000  0.000000            0.000000      0.000000   
14             0.0000  0.000000            0.000000      0.000000   
15             

#### Confirming the new size....

In [24]:
toronto_grouped.shape

(39, 235)

#### Let's find the 10 top venues for each Hood and then put that info on a DF

In [25]:
# funtion to return most common venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...    Light Rail Station   
3   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville           Pizza Place   
9                                    Davisville North  Gym / Fitness Center   
10                       Dufferin, Dovercourt Village              Pharmacy   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park                  Park   
13                           Garden District, Ryerson           Coffee Shop   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South                  Café   
16                     India Bazaar, The Beaches West        Sandwich Place   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   
20                        Moore Park, Summerhill East                  Park   
21                  North Toronto West, Lawrence Park        Clothing Store   
22                             Parkdale, Roncesvalles        Breakfast Spot   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn          Home Service   
28                                 Runnymede, Swansea           Coffee Shop   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown           Coffee Shop   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District           Coffee Shop   
33  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
34                The Annex, North Midtown, Yorkville                  Café   
35                                        The Beaches     Health Food Store   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

            2nd Most Common Venue 3rd Most Common Venue  \
0                          Bakery          Cocktail Bar   
1                  Breakfast Spot           Coffee Shop   
2            Gym / Fitness Center         Auto Workshop   
3                  Airport Lounge      Airport Terminal   
4                  Sandwich Place    Italian Restaurant   
5                            Café                  Park   
6                Sushi Restaurant   Japanese Restaurant   
7                      Restaurant                  Café   
8                    Dessert Shop        Sandwich Place   
9                           Hotel      Department Store   
10                         Bakery         Grocery Store   
11                           Café                 Hotel   
12                          Trail         Jewelry Store   
13                 Clothing Store                  Ca

### Last Part: Cluster Neighborhoods

#### Run K-means to cluster the hoods in 5 clusters

In [27]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_ 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 4, 0, 1, 0,
       0, 0, 0, 0, 1, 3, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0], dtype=int32)

#### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [28]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               2     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub                Bakery                  Park   
1      Sushi Restaurant           Yoga Studio             Nightclub   
2        Clothing Store                  Café                 Hotel   
3                  Café        Cosmetics Shop   American Restaurant   
4                 Trail                   Pub          Neighborhood   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0                  Café        Breakfast Spot                    Theater   
1         Smoothie Shop        Sandwich Place              Burrito Place   
2   Japanese Restaurant        Cosmetics Shop  Middle Eastern Restaurant   
3          Cocktail Bar                   Gym                      Hotel   
4           Yoga Studio      Doner Restaurant             Discount Store   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0           Yoga Studio           Event Space        Performing Arts Venue  
1            Restaurant                  Café           College Auditorium  
2       Bubble Tea Shop      Ramen Restaurant         Fast Food Restaurant  
3        Farmers Market    Italian Restaurant          Japanese Restaurant  
4   Distribution Center               Dog Run  Eastern European Restaurant

#### Lets' visualize the results

In [29]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Let's examine the clusters

#### Cluster 1: the most diverse of all, includes all types of venues, mostly focused on lot's of popular coffe shops

In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0              Pharmacy   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0        Sandwich Place   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0           Coffee Shop   
20   Central Toronto               0  Gym / Fitness Center   
22      West Toronto               0                  Café   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0                  Café   
25      West Toronto               0        Breakfast Spot   
26   Central Toronto               0           Pizza Place   
27  Downtown Toronto               0                  Café   
28      West Toronto               0           Coffee Shop   
30  Downtown Toronto               0                  Café   
31   Central Toronto               0           Coffee Shop   
32  Downtown Toronto               0       Airport Service   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub                Bakery   
1                Sushi Restaurant           Yoga Studio   
2                  Clothing Store                  Café   
3                            Café        Cosmetics Shop   
5                          Bakery          Cocktail Bar   
6                  Sandwich Place    Italian Restaurant   
7                            Café                  Park   
8                            Café            Restaurant   
9                          Bakery         Grocery Store   
10                       Aquarium                  Café   
11                    Men's Store      Asian Restaurant   
12                    Coffee Shop    Italian Restaurant   
13                          Hotel                  Café   
14                 Breakfast Spot           Coffee Shop   
15           Fast Food Restaurant           Pizza Place   
16                     Restaurant                  Café   
17            American Restaurant                Bakery   
20                          Hotel      Department Store   
22             Mexican Restaurant       Thai Restaurant   
23                    Coffee Shop           Yoga Studio   
24                 Sandwich Place           Coffee Shop   
25                      Gift Shop            Restaurant   
26                   Dessert Shop        Sandwich Place   
27                            Bar    Italian Restaurant   
28                           Café      Sushi Restaurant   
30  Vegetarian / Vegan Restaurant           Coffee Shop   
31            American Restaurant    Light Rail Station   
32                 Airport Lounge      Airport Terminal   
34             Italian Restaurant    Seafood Restaurant   
35                           Café                Bakery   
36                           Café                 Hotel   
37              

#### Cluster 2: mostly focused on active people who enjoys to practice sports on the open air

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
21   Central Toronto               1                  Park   
29   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                 Trail         Jewelry Store      Sushi Restaurant   
29            Playground          Tennis Court            Restaurant   
33            Playground                 Trail         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
21           Yoga Studio                        Diner           Event Space   
29     Electronics Store  Eastern European Restaurant           Escape Room   
33  Ethiopian Restaurant                  Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21         Ethiopian Restaurant           Escape Room      Electronics Store  
29                   Donut Shop         Deli / Bodega                Dog Run  
33  Eastern European Restaurant            Donut Shop       Doner Restaurant

#### Cluster 3: the only one with the most popular Health food store.

In [32]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  East Toronto               2     Health Food Store                 Trail   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                   Pub          Neighborhood           Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4      Doner Restaurant        Discount Store   Distribution Center   

  9th Most Common Venue       10th Most Common Venue  
4               Dog Run  Eastern European Restaurant

#### Cluster 4: The only one with a Home service and a Garden as 1st and 2nd most common venues

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               3          Home Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Yoga Studio          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
19           Escape Room     Electronics Store  Eastern European Restaurant

#### Cluster 5: the one with also a park as a most popular venue, but the rest are very distinct from the CLuster 2. Most focused on maybe older people?

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                Lawyer              Bus Line           Swim School   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18                 Diner           Event Space  Ethiopian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
18           Escape Room     Electronics Store  Eastern European Restaurant